File: sdk/python/src/fr_ir/flow_facts.py
"""Paged model facts and exact syntax links with independent disclosure coverage."""
from __future__ import annotations

from dataclasses import dataclass
import hashlib
import json
from pathlib import Path
import re
from typing import Any, Mapping

from .context import ObjectStore, restore_stored_value, store_merkle_value
from .origins import SemanticOrigins
from .runtime import FrClient, FrReport, FrRuntimeError, Occurrence

KINDS = {"return", "raise", "witness", "completion", "summary-return", "summary-raise",
         "summary-sink", "summary-completion", "boundary"}


def _digest(value: Any, *, canonical: bool = True) -> str:
    return hashlib.sha256(json.dumps(value, sort_keys=canonical, ensure_ascii=False,
                                     separators=(",", ":")).encode()).hexdigest()


def _identity(value: Any, prefix: str = "") -> bool:
    return isinstance(value, str) and re.fullmatch(re.escape(prefix) + r"[0-9a-f]{64}", value) is not None


def _read(client: FrClient, action: Any, route: str) -> FrReport:
    if (not isinstance(action, Mapping) or not isinstance(action.get("arguments"), list)
            or action["arguments"][:2] != ["project", route]
            or any(not isinstance(arg, str) for arg in action["arguments"])):
        raise FrRuntimeError("expected a read-only flow fact action")
    return client.call(*action["arguments"])


@dataclass(frozen=True)
class FlowFact:
    id: str
    basis: str
    core: Mapping[str, Any]
    follow: Mapping[str, Any] | None

    @property
    def kind(self) -> str:
        return self.core["kind"]

    @classmethod
    def from_data(cls, row: Any, basis: str, inputs: str, analysis_complete: bool) -> FlowFact:
        core = row["core"]
        kind = core["kind"]
        if (row["basis"] != basis or row["id"] != "frff1:" + _digest([basis, core])
                or kind not in KINDS or core["input_digest"] != inputs
                or core["rule"] != f"python-scalar-facts-1:{kind}"
                or core["confidence"] != ("observed-cutoff" if kind == "boundary" else "model-derived")
                or core["scope"] != ("symbolic-function" if kind.startswith("summary-") else "selected-entry")
                or type(core["analysis_complete"]) is not bool or core["analysis_complete"] != analysis_complete
                or type(core["evidence_count"]) is not int or core["evidence_count"] < 0
                or not _identity(core["evidence_digest"]) or not isinstance(core["conclusion"], dict)
                or not isinstance(core["function"], str) or not core["function"]
                or core["assumptions_ref"] != "/analysis/assumptions" or core["omissions_ref"] != "/analysis/omitted"):
            raise FrRuntimeError("flow fact has inconsistent identity or scope")
        return cls(row["id"], basis, core, row.get("follow"))


@dataclass(frozen=True)
class FactPoint:
    index: int
    occurrence: Occurrence
    rule: Mapping[str, Any]
    mapping: Mapping[str, Any]
    raw_occurrence: Mapping[str, Any]

    def semantic(self, client: FrClient, index: int = 0) -> SemanticOrigins:
        try:
            link = self.mapping["items"][index]
        except (IndexError, KeyError) as error:
            raise FrRuntimeError("semantic mapping was not disclosed") from error
        report = SemanticOrigins.from_report(_read(client, link["follow"], "semantic"))
        if (report.revision != self.occurrence.revision or report.basis != link["semantic_basis"]
                or not any(item.id == link["id"] for item in report.for_occurrence(self.occurrence))):
            raise FrRuntimeError("semantic link no longer matches this occurrence")
        return report


@dataclass(frozen=True)
class FactCollection:
    facts: tuple[FlowFact, ...]
    evidence: tuple[FactPoint, ...]
    complete: bool
    disclosure_complete: bool
    pages: int
    continuation: Mapping[str, Any] | None


@dataclass(frozen=True)
class FlowFacts:
    report: FrReport
    items: tuple[FlowFact, ...]
    fact: FlowFact | None
    evidence: tuple[FactPoint, ...]

    @property
    def complete(self) -> bool:
        return self.report.at("/complete")

    @classmethod
    def inspect(cls, client: FrClient, target: str, *, rules: Path | None = None,
                context: str = "generic", steps: int = 1024, depth: int = 8, limit: int = 16,
                evidence_limit: int = 8, max_bytes: int = 65_536) -> FlowFacts:
        arguments = ["flow-facts", target, "--context", context, "--steps", str(steps),
                     "--depth", str(depth), "--limit", str(limit), "--evidence-limit", str(evidence_limit),
                     "--bytes", str(max_bytes)]
        if rules is not None:
            arguments.extend(["--rules", str(rules)])
        return cls.from_report(client.project(*arguments))

    @classmethod
    def from_report(cls, report: FrReport) -> FlowFacts:
        try:
            data = report.to_data()
            analysis, basis, revision = data["analysis"], data["basis"], data["revision"]
            if (report.schema != "fr-flow-facts-1" or data["mutation_authority"] is not False
                    or not _identity(basis, "frfb1:") or not _identity(revision)
                    or not _identity(analysis["input_digest"]) or type(analysis["complete"]) is not bool
                    or basis != "frfb1:" + _digest([revision, analysis["input_digest"], data["analyzer"]])
                    or analysis["input_digest"] != _digest(analysis["inputs"])
                    or not isinstance(analysis["cutoffs"], list) or not isinstance(analysis["omitted"], dict)
                    or analysis["complete"] and (analysis["cutoffs"] or analysis["omitted"])):
                raise FrRuntimeError("flow fact report has inconsistent analysis identity")
            parse = lambda row: FlowFact.from_data(row, basis, analysis["input_digest"], analysis["complete"])
            items = tuple(parse(row) for row in data["items"])
            fact = None if data["fact"] is None else parse(data["fact"])
            if len({item.id for item in items}) != len(items):
                raise FrRuntimeError("duplicate flow fact")
            evidence = []
            for row in data["evidence"]:
                raw = row["occurrence"]
                point = Occurrence.from_data(raw)
                if (point.revision != revision or point.id != "fro1:" + _digest([
                        revision, point.path, {"start":point.location.span.start,"end":point.location.span.end}, point.role], canonical=False)
                        or row["rule"]["id"] != f"python-scalar-facts-1:occurrence:{point.role}"):
                    raise FrRuntimeError("fact occurrence has inconsistent identity")
                mapping = row["mapping"]
                if (mapping["status"] not in {"mapped", "absent", "incomplete", "unavailable"}
                        or type(mapping["complete"]) is not bool or not isinstance(mapping["items"], list)
                        or len(mapping["items"]) > 4 or (mapping["status"] == "mapped") != bool(mapping["items"])
                        or mapping["status"] == "absent" and not mapping["complete"]
                        or mapping["status"] in {"incomplete", "unavailable"} and mapping["complete"]):
                    raise FrRuntimeError("semantic mapping has inconsistent coverage")
                for link in mapping["items"]:
                    pointer, semantic = link["pointer"], link["semantic_basis"]
                    if (not _identity(semantic, "frsm1:") or not isinstance(pointer, str)
                            or not pointer.startswith("/model/") or link["id"] != "frso1:" + _digest([semantic, pointer])
                            or link["relation"] not in {"exact", "member-of-multiple"}):
                        raise FrRuntimeError("semantic link has inconsistent identity")
                    prefix = "/model/items/0/value/body"
                    if link["body_pointer"] is not None and (
                            not _identity(link["body_basis"], "frsb1:") or not pointer.startswith(prefix + "/")
                            or link["body_pointer"] != "/body" + pointer[len(prefix):]):
                        raise FrRuntimeError("semantic link has an invalid authoring pointer")
                evidence.append(FactPoint(row["index"], point, row["rule"], mapping, raw))
            page = data["page"]
            if any(type(page[key]) is not int or page[key] < 0 for key in ["before", "returned", "remaining", "total"]):
                raise FrRuntimeError("invalid fact page counts")
            points = evidence if fact is not None else items
            if (page["returned"] != len(points) or page["total"] != page["before"] + len(points) + page["remaining"]
                    or (page["next"] is None) != (page["remaining"] == 0)
                    or (data["continuation"] is None) != (page["next"] is None)
                    or fact is not None and (items or page["total"] != fact.core["evidence_count"])
                    or fact is None and evidence):
                raise FrRuntimeError("inconsistent fact page coverage")
            if any(type(point.index) is not int or point.index != page["before"] + index for index, point in enumerate(evidence)):
                raise FrRuntimeError("fact evidence has a gap or reordered index")
            disclosed = page["before"] == 0 and page["remaining"] == 0
            if (type(data["complete"]) is not bool or type(data["disclosure"]["complete"]) is not bool
                    or data["disclosure"]["complete"] != disclosed or data["complete"] != (analysis["complete"] and disclosed)
                    or data["disclosure"]["kind"] != ("fact-evidence" if fact else "fact-catalogue")):
                raise FrRuntimeError("fact completeness differs from analysis or disclosure coverage")
            if fact is not None and disclosed and _digest([point.raw_occurrence for point in evidence]) != fact.core["evidence_digest"]:
                raise FrRuntimeError("fact evidence digest mismatch")
            return cls(report, items, fact, tuple(evidence))
        except (KeyError, TypeError, ValueError, AttributeError) as error:
            raise FrRuntimeError("malformed flow fact report") from error

    def _same(self, result: FlowFacts) -> None:
        for pointer in ["/basis", "/revision", "/target", "/analysis/input_digest"]:
            if result.report.at(pointer) != self.report.at(pointer):
                raise FrRuntimeError("fact follow crossed an analysis identity")

    def explain(self, client: FrClient, fact: FlowFact) -> FlowFacts:
        if fact.basis != self.report.at("/basis"):
            raise FrRuntimeError("fact belongs to another analysis")
        result = self.from_report(_read(client, fact.follow, "flow-facts"))
        self._same(result)
        if result.fact is None or result.fact.id != fact.id or result.report.at("/page/before") != 0:
            raise FrRuntimeError("fact follow returned a different explanation")
        return result

    def next(self, client: FrClient) -> FlowFacts:
        action = self.report.at("/continuation")
        if action is None:
            raise FrRuntimeError("fact page has no continuation")
        result = self.from_report(_read(client, action, "flow-facts"))
        self._same(result)
        if ((None if self.fact is None else self.fact.id) != (None if result.fact is None else result.fact.id)
                or result.report.at("/page/before") != self.report.at("/page/before") + self.report.at("/page/returned")):
            raise FrRuntimeError("fact continuation skipped or repeated evidence")
        return result

    def collect(self, client: FrClient, *, max_pages: int = 64) -> FactCollection:
        if type(max_pages) is not int or not 1 <= max_pages <= 64:
            raise FrRuntimeError("fact collection needs a page budget between 1 and 64")
        current, items, evidence, pages = self, list(self.items), list(self.evidence), 1
        while current.report.at("/continuation") is not None and pages < max_pages:
            current = current.next(client)
            items.extend(current.items)
            evidence.extend(current.evidence)
            pages += 1
        disclosed = self.report.at("/page/before") == 0 and current.report.at("/continuation") is None
        if len({item.id for item in items}) != len(items):
            raise FrRuntimeError("fact collection repeats identities")
        if self.fact is not None and disclosed and _digest([point.raw_occurrence for point in evidence]) != self.fact.core["evidence_digest"]:
            raise FrRuntimeError("collected fact evidence digest mismatch")
        return FactCollection(tuple(items), tuple(evidence), disclosed and self.report.at("/analysis/complete"),
                              disclosed, pages, current.report.at("/continuation"))

    def persist(self, store: ObjectStore) -> str:
        return store_merkle_value(store, self.report.to_data()).digest

    @classmethod
    def restore(cls, store: ObjectStore, digest: str) -> FlowFacts:
        return cls.from_report(FrReport(restore_stored_value(store, digest), ()))

File: src/project/flow_facts.rs
use super::{dataflow, flow_fact_origins::Linker, hash, occurrence::Occurrence, page, Project};
use anyhow::{ensure, Context, Result};
use clap::Args;
use serde_json::{json, Value};
use std::path::PathBuf;

const RULE: &str = "python-scalar-facts-1";

pub fn fact_catalogue_complete(analysis: bool, from_start: bool, no_remaining: bool) -> bool {
    analysis && from_start && no_remaining
}

#[derive(Args)]
pub struct Options {
    target: String,
    #[arg(long)]
    rules: Option<PathBuf>,
    #[arg(long, default_value = "generic")]
    context: String,
    #[arg(long, default_value_t = 1024)]
    steps: usize,
    #[arg(long, default_value_t = 8)]
    depth: usize,
    #[arg(long, default_value_t = 16)]
    limit: usize,
    #[arg(long, conflicts_with = "fact")]
    cursor: Option<String>,
    #[arg(long)]
    fact: Option<String>,
    #[arg(long, default_value_t = 8)]
    evidence_limit: usize,
    #[arg(long, requires = "fact")]
    evidence_cursor: Option<String>,
    #[arg(long, default_value_t = 65536)]
    bytes: usize,
}

struct Fact {
    row: Value,
    evidence: Vec<Occurrence>,
}

fn add(
    facts: &mut Vec<Fact>,
    report: &Value,
    basis: &str,
    kind: &str,
    function: &str,
    conclusion: Value,
    evidence: Value,
) -> Result<()> {
    let evidence: Vec<Occurrence> = serde_json::from_value(evidence)?;
    let core = json!({"kind":kind,"function":function,"conclusion":conclusion,
        "rule":format!("{RULE}:{kind}"),"input_digest":report["input_digest"],
        "confidence":if kind == "boundary" {"observed-cutoff"} else {"model-derived"},
        "scope":if kind.starts_with("summary-") {"symbolic-function"} else {"selected-entry"},
        "analysis_complete":report["complete"],"assumptions_ref":"/analysis/assumptions",
        "omissions_ref":"/analysis/omitted","evidence_count":evidence.len(),"evidence_digest":hash(serde_json::to_value(&evidence)?)?});
    let id = format!("frff1:{}", hash((basis, &core))?);
    facts.push(Fact {
        row: json!({"id":id,"basis":basis,"core":core}),
        evidence,
    });
    Ok(())
}

fn facts(report: &Value, basis: &str) -> Result<Vec<Fact>> {
    let mut rows = Vec::new();
    for kind in ["returns", "exceptional_returns"] {
        if let Some(flows) = report[kind].as_object() {
            for (origin, trace) in flows {
                add(
                    &mut rows,
                    report,
                    basis,
                    if kind == "returns" { "return" } else { "raise" },
                    "entry",
                    json!({"origin":origin}),
                    trace["occurrences"].clone(),
                )?;
            }
        }
    }
    if let Some(witnesses) = report["witnesses"].as_array() {
        for witness in witnesses {
            add(
                &mut rows,
                report,
                basis,
                "witness",
                "entry",
                json!({"origin":witness["trace"]["origin"],
                "sink":witness["sink"],"site":witness["site"]["id"],"context":witness["context"]}),
                witness["trace"]["occurrences"].clone(),
            )?;
        }
    }
    if report["completion"].is_object() {
        add(
            &mut rows,
            report,
            basis,
            "completion",
            "entry",
            report["completion"].clone(),
            json!([]),
        )?;
    }
    if let Some(functions) = report["function_summaries"]["functions"].as_object() {
        for (name, summary) in functions {
            for (field, kind) in [
                ("returns", "summary-return"),
                ("exceptional_returns", "summary-raise"),
            ] {
                for (origin, trace) in summary[field].as_object().into_iter().flatten() {
                    add(
                        &mut rows,
                        report,
                        basis,
                        kind,
                        name,
                        json!({"origin":origin}),
                        trace["occurrences"].clone(),
                    )?;
                }
            }
            for effect in summary["sinks"]
                .as_object()
                .into_iter()
                .flat_map(|object| object.values())
            {
                for (origin, trace) in effect["flow"].as_object().into_iter().flatten() {
                    add(
                        &mut rows,
                        report,
                        basis,
                        "summary-sink",
                        name,
                        json!({"origin":origin,"sink":effect["sink"],
                        "site":effect["site"]["id"],"context":report["context"]}),
                        trace["occurrences"].clone(),
                    )?;
                }
            }
            add(
                &mut rows,
                report,
                basis,
                "summary-completion",
                name,
                json!({"normal_return":summary["normal_return"],"may_raise":summary["may_raise"],
                    "evaluations":summary["evaluations"],"converged":report["function_summaries"]["converged"]}),
                json!([]),
            )?;
        }
    }
    for reason in report["cutoffs"].as_array().into_iter().flatten() {
        add(
            &mut rows,
            report,
            basis,
            "boundary",
            "analysis",
            json!({"reason":reason}),
            json!([]),
        )?;
    }
    Ok(rows)
}

impl Options {
    fn arguments(&self) -> Vec<String> {
        let mut args = vec![
            "project".into(),
            "flow-facts".into(),
            self.target.clone(),
            "--steps".into(),
            self.steps.to_string(),
            "--depth".into(),
            self.depth.to_string(),
            "--context".into(),
            self.context.clone(),
            "--limit".into(),
            self.limit.to_string(),
            "--evidence-limit".into(),
            self.evidence_limit.to_string(),
            "--bytes".into(),
            self.bytes.to_string(),
        ];
        if let Some(rules) = &self.rules {
            args.extend(["--rules".into(), rules.to_string_lossy().into_owned()]);
        }
        args
    }

    fn action(&self, fact: Option<&str>, cursor: Option<&str>) -> Value {
        let mut arguments = self.arguments();
        if let Some(id) = fact {
            arguments.extend(["--fact".into(), id.into()]);
        }
        if let Some(cursor) = cursor {
            arguments.extend([
                if fact.is_some() {
                    "--evidence-cursor"
                } else {
                    "--cursor"
                }
                .into(),
                cursor.into(),
            ]);
        }
        json!({"arguments":arguments})
    }
}

impl Project<'_> {
    pub(super) fn flow_facts(&self, options: &Options) -> Result<Value> {
        ensure!(
            (1..=64).contains(&options.limit) && (1..=64).contains(&options.evidence_limit),
            "fact and evidence limits must be between 1 and 64."
        );
        ensure!(
            (4096..=1_048_576).contains(&options.bytes),
            "fact response budget must be between 4096 and 1048576."
        );
        let analysis = self.dataflow(&dataflow::Options::for_facts(
            &options.target,
            options.rules.clone(),
            &options.context,
            options.steps,
            options.depth,
        ))?;
        let analyzer = hash((
            RULE,
            include_str!("flow_facts.rs"),
            include_str!("flow_fact_origins.rs"),
        ))?;
        let basis = format!(
            "frfb1:{}",
            hash((&self.revision, &analysis["input_digest"], &analyzer))?
        );
        let rows = facts(&analysis, &basis)?;
        let mut result = json!({"schema":"fr-flow-facts-1","revision":self.revision,"basis":basis,
            "handle_prefix":analysis["handle_prefix"],"coverage":analysis["coverage"],
            "target":options.target,"analyzer":analyzer,"rule_version":RULE,
            "analysis":{"input_digest":analysis["input_digest"],"inputs":analysis["inputs"],"semantics":analysis["semantics"],
                "complete":analysis["complete"],"cutoffs":analysis["cutoffs"],"omitted":analysis.get("omitted").cloned().unwrap_or(json!({})),
                "assumptions":analysis["assumptions"],"rules_digest":analysis["rules_digest"],"context":analysis["context"],
                "execution":analysis["execution"]},
            "claim":"model-derived value propagation; neither executable paths nor source implementation proofs.",
            "mutation_authority":false,"items":[],"evidence":[],"fact":Value::Null,
            "disclosure":{"kind":if options.fact.is_some() {"fact-evidence"} else {"fact-catalogue"}},
            "budget":{"response_bytes":options.bytes,"semantic_queries":8},
            "continuation":Value::Null});
        let key = format!("frff-page:{}", hash((&basis, &options.fact))?);
        let mut linker = Linker::new(self);
        let (total, limit, cursor) = if let Some(id) = &options.fact {
            let fact = rows
                .iter()
                .find(|row| row.row["id"] == *id)
                .context("unknown or stale flow fact; restart the fact catalogue.")?;
            result["fact"] = fact.row.clone();
            result["fact"]["follow"] = options.action(Some(id), None);
            (
                fact.evidence.len(),
                options.evidence_limit,
                options.evidence_cursor.as_deref(),
            )
        } else {
            (rows.len(), options.limit, options.cursor.as_deref())
        };
        let (start, end, _) = page(total, limit, cursor, &key)?;
        let field = if options.fact.is_some() {
            "evidence"
        } else {
            "items"
        };
        for index in start..end {
            let item = if let Some(id) = &options.fact {
                let fact = rows.iter().find(|row| row.row["id"] == *id).unwrap();
                let point = &fact.evidence[index];
                let mut item = linker.explain(point)?;
                item["index"] = json!(index);
                item["rule"] = occurrence_rule(&point.role, &analysis["rules_digest"]);
                item
            } else {
                let mut row = rows[index].row.clone();
                row["follow"] = options.action(row["id"].as_str(), None);
                row
            };
            result[field].as_array_mut().unwrap().push(item);
        }
        loop {
            let returned = result[field].as_array().unwrap().len();
            let end = start + returned;
            let next = (end < total).then(|| format!("{key}:{end}"));
            result["page"] = json!({"before":start,"returned":returned,"remaining":total-end,"total":total,"next":next});
            result["continuation"] = next
                .as_deref()
                .map(|cursor| options.action(options.fact.as_deref(), Some(cursor)))
                .unwrap_or(Value::Null);
            result["disclosure"]["complete"] = json!(start == 0 && end == total);
            result["complete"] = json!(fact_catalogue_complete(
                analysis["complete"] == true,
                start == 0,
                end == total
            ));
            result["semantic_queries"] = json!(linker.queries());
            if serde_json::to_vec(&result)?.len() + 512 <= options.bytes {
                break;
            }
            ensure!(
                returned > 0,
                "fact metadata exceeds response budget; increase --bytes."
            );
            result[field].as_array_mut().unwrap().pop();
            ensure!(
                returned > 1 || total == 0,
                "one fact or explanation exceeds response budget; increase --bytes."
            );
        }
        Ok(result)
    }
}

fn occurrence_rule(role: &str, external_digest: &Value) -> Value {
    let contract = matches!(
        role,
        "source" | "sink" | "propagation-rule" | "sanitizer-context-mismatch"
    );
    let claim = match role {
        "source" => "The caller supplied this source contract.",
        "sink" => "The caller supplied this sink contract.",
        "parameter" | "entry-parameter" | "summary-parameter" => {
            "This occurrence binds a positional parameter."
        }
        "definition" => "This assignment replaces prior scalar origins.",
        "use" => "This use reads the current abstract scalar value.",
        "return" => "This explicit return carries the recorded origins.",
        "raise" | "summary-raise" => {
            "This explicit exceptional effect carries the recorded origins."
        }
        "call-result" | "summary-call" => {
            "This call substitutes arguments into a symbolic function summary."
        }
        "expression" => "This expression combines explicit operand origins.",
        "sanitizer-context-mismatch" => {
            "The sanitizer contract does not apply in the selected context."
        }
        "propagation-rule" => "The caller supplied this propagation contract.",
        _ => "This is a recorded syntax occurrence in a model derivation.",
    };
    json!({"id":format!("{RULE}:occurrence:{role}"),"claim":claim,
        "external_rules_digest":if contract {external_digest.clone()} else {Value::Null},
        "relation":"derivation membership; no feasible-path or per-edge proof"})
}
